### PR TITLE
OSSM-1421: Fixed leaking BIGNUMs

### PR DESCRIPTION
**Pull request by tedjpoole**
In `createEcKeyFromJwkEC()` the two `BIGNUM` objects were leaking because `EC_KEY_set_public_key_affine_coordinates()` does _not_ take ownership of them. They would also leak in either of the error cases.

In `createRsaFromJwk()` the two `BIGNUM` objects would leak in either of the error cases. Note that `RSA_set0_key()` _does_ however take ownership of the two `BIGNUM` objects.

### CODE REVIEW

File: repositories.bzl [hunk 1]
```diff
@@ -127,8 +127,8 @@ def _rules_proto_repositories():
         urls = ["https://github.com/bazelbuild/rules_proto/archive/" + RULES_PROTO_COMMIT + ".tar.gz"],
     )
 
-ZLIB_RELEASE = "1.2.11"
-ZLIB_SHA256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1"
+ZLIB_RELEASE = "1.2.12"
```
Comment (jwendell) [inline repositories.bzl:130]: Is this related?

File: src/jwks.cc [hunk 2]
```diff
@@ -49,11 +49,6 @@ inline const uint8_t* castToUChar(const std::string& str) {
 }
 
 /** Class to create key object from string of public key, formatted in PEM
-inline const char* castToChar(const std::string& str) {
```
Comment (jwendell) [inline src/jwks.cc:52]: I'm guessing this is not used anywhere?

**Review by jwendell — COMMENTED** [review]
Overall, LGTM.

Does this apply to upstream too? If so, please submit a PR there https://github.com/google/jwt_verify_lib

File: src/jwks.cc [hunk 3]
```diff
@@ -85,43 +80,41 @@ class KeyGetter : public WithStatus {
...
   }
 
   bssl::UniquePtr<RSA> createRsaFromJwk(const std::string& n,
-                                          const std::string& e) {
+                                        const std::string& e) {
     bssl::UniquePtr<RSA> rsa(RSA_new());
-    BIGNUM *bn_n = createBigNumFromBase64UrlString(n);
-    BIGNUM *bn_e = createBigNumFromBase64UrlString(e);
+    bssl::UniquePtr<BIGNUM> bn_n = createBigNumFromBase64UrlString(n);
+    bssl::UniquePtr<BIGNUM> bn_e = createBigNumFromBase64UrlString(e);
```
Comment (oschaaf) [inline src/jwks.cc:104]: If either `bn_n` or `bn_e` ends up being equal to `nullptr` at this point, I think the docs to `RSA_set0_key` say we mustn't call it [1].  I think this may only happen when `absl::WebSafeBase64Unescape` failed in `createBigNumFromBase64UrlString()`. If that can't happen, we should propose a doc comment to make it explicit that createRsaFromJwk trusts its inputs won't fail unescaping. If it can happen, we need to handle this case?

[1] https://www.openssl.org/docs/man1.1.1/man3/RSA_set0_key.html says

```
The n, e and d parameter values can be set by calling RSA_set0_key() and passing the new values for n, e and d as parameters to the function. The values n and e must be non-NULL the first time this function is called on a given RSA object. 
```

**Review by oschaaf — CHANGES_REQUESTED** [review]

File: repositories.bzl (hunk 1 above)
Comment (tedjpoole) [reply repositories.bzl:130]: No it's not directly related to the fix. However, the zlib 1.2.11 archive is no longer available for download. I had to update it to 1.2.12 so I could build & test.

File: src/jwks.cc (hunk 2 above)
Comment (tedjpoole) [reply src/jwks.cc:52]: It looked like it had been badly merged into the middle of a comment block - so definitely not being used! There's an almost identical function just above here, which is used.

File: src/jwks.cc (hunk 3 above)
Comment (tedjpoole) [reply src/jwks.cc:104]: There is already a check for either `bn_n` or `bn_e` being `nullptr`, which will prevent `RSA_set0_key()` being called in that case.

File: src/jwks.cc [hunk 4]
```diff
@@ -85,43 +80,41 @@ class KeyGetter : public WithStatus {
...
 
     if (bn_n == nullptr || bn_e == nullptr) {
```
Comment (jwendell) [inline src/jwks.cc:105]: do you need to check using `.get()` here?

File: src/jwks.cc (hunk 3 above)
Comment (oschaaf) [reply src/jwks.cc:104]: Ah ok lgtm

**Review by oschaaf — APPROVED** [review]

File: src/jwks.cc (hunk 4 above)
Comment (tedjpoole) [reply src/jwks.cc:105]: No, there is an overloaded `==` operator that specifically takes `unique_ptr` & `nullptr` arguments, for use like this.

Comment (tedjpoole) [reply src/jwks.cc:105]: I considered changing it to use the `bool` operator (`!bn_n || !bn_e`) but using the `==` operator keeps it the same as upstream
